File: app/src/main/java/com/werebug/anmapwrapper/NmapScan.java
package com.werebug.anmapwrapper;

import android.os.Handler;
import android.util.Log;
import android.widget.Toast;

import java.io.IOException;
import java.io.InputStream;
import java.lang.ref.WeakReference;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Map;

public class NmapScan implements Runnable {
    private final WeakReference<MainActivity> mainActivityRef;
    private final Handler mainThreadHandler;
    private final ArrayList<String> command;
    private final String libDir;
    private boolean stopped = false;

    NmapScan(WeakReference<MainActivity> mainActivityRef,
             ArrayList<String> command,
             Handler mainThreadHandler,
             String libDir) {
        this.mainActivityRef = mainActivityRef;
        this.command = command;
        this.mainThreadHandler = mainThreadHandler;
        this.libDir = libDir;
    }

    @Override
    public void run() {
        ProcessBuilder processBuilder = new ProcessBuilder(command);
        Map<String, String> processEnv = processBuilder.environment();
        processEnv.put("LD_LIBRARY_PATH", libDir);
        processBuilder.redirectErrorStream(true);
        try {
            Process process = processBuilder.start();
            mainThreadHandler.post(() -> mainActivityRef.get().initScanView());
            InputStream processStdout = process.getInputStream();
            boolean exited = false;
            while (!exited && !stopped) {
                int outputByteCount = processStdout.available();
                if (outputByteCount > 0) {
                    byte[] bytes = new byte[outputByteCount];
                    processStdout.read(bytes);
                    mainThreadHandler.post(
                            () -> safeUpdateOutputView(
                                    new String(bytes, StandardCharsets.UTF_8), false)
                    );
                }
                if (stopped) {
                    process.destroy();
                }
                try {
                    process.exitValue();
                    exited = true;
                } catch (IllegalThreadStateException ignored) {}
            }
            if (stopped) {
                mainThreadHandler.post(
                        () -> Toast.makeText(
                                mainActivityRef.get(), "Stopped.", Toast.LENGTH_SHORT).show());
            }
            mainThreadHandler.post(() -> safeUpdateOutputView("", true));
        } catch (IOException e) {
            Log.e(MainActivity.LOG_TAG, e.getMessage());
            mainThreadHandler.post(
                    () -> Toast.makeText(
                            mainActivityRef.get(), e.getMessage(), Toast.LENGTH_LONG).show());
        }
    }

    private void safeUpdateOutputView(String message, boolean finished) {
        MainActivity mainActivity = mainActivityRef.get();
        if (mainActivity != null) {
            mainActivity.updateOutputView(message, finished);
        }
    }

    public void stopScan() {
        stopped = true;
    }
}

File: app/src/main/java/com/werebug/anmapwrapper/NmapScan.kt
package com.werebug.anmapwrapper

import android.os.Handler
import android.util.Log
import java.io.IOException
import java.nio.charset.StandardCharsets

class NmapScan(
  private val command: List<String>,
  private val mainHandler: Handler,
  private val listener: Listener
) : Runnable {

  enum class Outcome { COMPLETED, STOPPED, FAILED }

  interface Listener {
    fun onChunk(chunk: String)
    fun onError(message: String)
    fun onFinished(outcome: Outcome)
  }

  @Volatile
  private var stopped = false

  @Volatile
  private var process: Process? = null

  override fun run() {
    val processBuilder = ProcessBuilder(command)
    processBuilder.redirectErrorStream(true)
    val startedProcess: Process
    try {
      startedProcess = processBuilder.start()
    } catch (e: IOException) {
      val message = e.message ?: e.toString()
      Log.e(MainActivity.LOG_TAG, message)
      mainHandler.post {
        listener.onError(message)
        listener.onFinished(Outcome.FAILED)
      }
      return
    }
    process = startedProcess
    val processStdout = startedProcess.inputStream
    var readFailed = false
    var exitValue = -1
    try {
      val buffer = ByteArray(4096)
      while (true) {
        val bytesRead = processStdout.read(buffer)
        if (bytesRead <= 0) break
        val chunk = String(buffer, 0, bytesRead, StandardCharsets.UTF_8)
        mainHandler.post { listener.onChunk(chunk) }
      }
    } catch (e: IOException) {
      if (!stopped) {
        readFailed = true
        val message = e.message ?: e.toString()
        Log.e(MainActivity.LOG_TAG, message)
        mainHandler.post { listener.onError(message) }
      }
    } finally {
      try {
        processStdout.close()
      } catch (_: IOException) {
      }
      try {
        exitValue = startedProcess.waitFor()
      } catch (_: InterruptedException) {
        Thread.currentThread().interrupt()
      }
    }
    val outcome = when {
      stopped -> Outcome.STOPPED
      readFailed || exitValue != 0 -> Outcome.FAILED
      else -> Outcome.COMPLETED
    }
    mainHandler.post { listener.onFinished(outcome) }
  }

  fun stopScan() {
    stopped = true
    process?.destroy()
  }
}
